src/iterative_ensemble_smoother/sies.py

```diff
@@ -0,0 +1,416 @@
+               truncation factor `truncation`.
+
+    truncation : float
+        How much of the total energy (singular values squared) to keep in the
+        SVD when `inversion` equals `subspace_projected`. Choosing 1.0
+        retains the most information, while 0.0 removes all information.
+    seed : Union[None, int, np.random._generator.Generator], optional
+        Integer used to seed the random number generator. The default is None.
+    """
+
+    inversion_funcs = {
+        "direct": inversion_direct,
+        "subspace_exact": inversion_subspace_exact,
+        "subspace_projected": inversion_subspace_projected,
+    }
+
+    def __init__(
+        self,
+        parameters: npt.NDArray[np.double],
+        covariance: npt.NDArray[np.double],
+        observations: npt.NDArray[np.double],
+        *,
+        inversion: str = "subspace_exact",
+        truncation: float = 1.0,
+        seed: Union[None, int, np.random._generator.Generator] = None,
+    ):
+        _validate_inputs(
+            parameters=parameters, covariance=covariance, observations=observations
+        )
+
+        self.rng = np.random.default_rng(seed)
+        self.inversion = self.inversion_funcs[inversion]
+        self.truncation = truncation
+        self.X = parameters
+        self.d = observations
+        self.C_dd = covariance
+        self.A = (self.X - self.X.mean(axis=1, keepdims=True)) / np.sqrt(
+            self.X.shape[1] - 1
+        )
+
+        # Create and store the cholesky factorization of C_dd
+        if self.C_dd.ndim == 2:
+            # With lower=True, the cholesky factorization obeys:
+            # C_dd_cholesky @ C_dd_cholesky.T = C_dd
+            self.C_dd_cholesky = sp.linalg.cholesky(
+                self.C_dd,
+                lower=True,
+                overwrite_a=False,
+                check_finite=True,
+            )
+        else:
+            self.C_dd_cholesky = np.sqrt(self.C_dd)
+
+        # Equation (14)
+        self.D = (
+            self.d.reshape(-1, 1)
+            + sample_mvnormal(
+                C_dd_cholesky=self.C_dd_cholesky, rng=self.rng, size=self.X.shape[1]
+            )
+        ).astype(parameters.dtype, copy=False)
+
+        self.W = np.zeros(
+            shape=(self.X.shape[1], self.X.shape[1]), dtype=parameters.dtype
+        )
+
+    def evaluate_objective(
+        self, W: npt.NDArray[np.double], Y: npt.NDArray[np.double]
+    ) -> npt.NDArray[np.double]:
+        """Evaluate the objective function in Equation (18), taking the mean
+        over each ensemble member.
+
+        Parameters
+        ----------
+        W : npt.NDArray[np.double]
+            Current weight matrix W, which represents the current X_i as a linear
+            combination of the prior. See equation (17) and line 9 in Algorithm 1.
+        Y : npt.NDArray[np.double]
+            Reponses when evaluating the model at X_i. In other words, Y = g(X_i).
```

Comment:
Evensen writes in equation (1) that y=g(x).
However, in equation (30) he writes Y_i = g(X_i)(I - 1/N11^T)/sqrt(N-1) which means that Y is centered.
Should we use a different symbol for non-centered and centered responses?

Reply:
write extra sentence to clarify this. but leave it as Y unless we figure out something better.